server/src/main/java/org/elasticsearch/index/query/ExistsQueryBuilder.java

```diff
@@ -181,6 +192,45 @@ private static Query newObjectFieldExistsQuery(QueryShardContext context, String
         return new ConstantScoreQuery(booleanQuery.build());
     }
 
+    /**
+     * Helper method to validate whether field is mapped
+     * @return true if the field is mapped, false otherwise
+     */
+
+    private static boolean isFieldMapped(QueryShardContext context, String fieldPattern) {
+        final FieldNamesFieldMapper.FieldNamesFieldType fieldNamesFieldType = (FieldNamesFieldMapper.FieldNamesFieldType) context
+            .getMapperService().fieldType(FieldNamesFieldMapper.NAME);
+
+        if (fieldNamesFieldType == null) {
+            // can only happen when no types exist, so no docs exist either
+            return false;
+        }
+
+        final Collection<String> fields;
+        if (context.getObjectMapper(fieldPattern) != null) {
+            // the _field_names field also indexes objects, so we don't have to
+            // do any more work to support exists queries on whole objects
+            fields = Collections.singleton(fieldPattern);
+        } else {
+            fields = context.simpleMatchToIndexNames(fieldPattern);
+        }
+
+        if (fields.size() == 1) {
+            String field = fields.iterator().next();
+            MappedFieldType fieldType = context.getMapperService().fieldType(field);
+            if (fieldType == null) {
+                // The field does not exist as a leaf but could be an object so
+                // check for an object mapper
+                if (context.getObjectMapper(field) != null) {
+                    return false;
+                }
+                return false; // no field mapped
+            }
+        }
+
+        return true;
+    }
```

Comment:
It's very easy for this method to get out-of-sync with `newFilter`, which would likely translate into bugs. Can you make this method called by `newFilter`? For instance I wonder that it could return a collection of fields instead of a boolean, and all above `return false` statements would become `return Collections.emptySet();` and then in `doRewrite` we'd check whether the return collection is empty? Finally in `doToQuery` we could raise an exception when the returned collection is empty, as it would mean that the query wasn't rewritten first even though it is a requirement of the API.

Reply:
A good suggestion 👍 I made the necessary changes. Thanks for pointing out.